docs/advanced/classes.rst

```diff
@@ -559,6 +559,39 @@ crucial that instances are deallocated on the C++ side to avoid memory leaks.
     py::class_<MyClass, std::unique_ptr<MyClass, py::nodelete>>(m, "MyClass")
         .def(py::init<>())
 
+.. _destructors_that_call_python:
+
+Destructors that call Python
+============================
+
+If a Python function is invoked from a C++ destructor, an exception may be thrown
+of type :class:`error_already_set`. If this error is thrown out of a class destructor,
+``std::terminate()`` will be called, terminating the process. Class destructors
+must catch all exceptions of type :class:`error_already_set` to discard the Python
+exception using :meth:`error_already_set::discard_as_unraisable`. (In addition,
```

Comment:
Do you mean `std::terminate()` -> `std::terminate`?

I think classes.rst needs a bigger overhaul than this but was trying to limit the scope of changes for now. In particular, `py::error_already_set` serves a different function from the C++ exception wrappers, and needs to be separately explained separately from them and then an example added that brings the two concepts together.

We could add `py::builtin_exception::discard_as_unraisable`. It's a little trickier to convert a `std::exception` into a Python exception, because I don't think we should invoke the exception translator as part of dealing with an unraisable exception.

Reply:
> Do you mean `std::terminate()` -> `std::terminate`?

No, sorry, I meant the parentheses around "In addition, ...".

> I think classes.rst needs a bigger overhaul than this but was trying to limit the scope of changes for now. In particular, `py::error_already_set` serves a different function from the C++ exception wrappers, and needs to be separately explained separately from them and then an example added that brings the two concepts together.
> 
> We could add `py::builtin_exception::discard_as_unraisable`. It's a little trickier to convert a `std::exception` into a Python exception, because I don't think we should invoke the exception translator as part of dealing with an unraisable exception.

Alright, yes, I can see. In that case, this is definitely better than what we had, so more changes can come later! :-)